File: backend/src/main/java/com/gxs/bank/model/Card.java
package com.gxs.bank.model;

import com.fasterxml.jackson.annotation.JsonIgnore;
import jakarta.persistence.*;
import lombok.*;
import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.UUID;

@Entity
@Table(name = "cards")
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Card {
    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", nullable = false)
    @JsonIgnore
    private User user;

    @Column(name = "card_number_last4", nullable = false, length = 4)
    private String cardNumberLast4;

    @Column(name = "card_holder_name")
    private String cardHolderName;

    @Enumerated(EnumType.STRING)
    @Column(name = "card_type", nullable = false)
    private CardType cardType;

    @Enumerated(EnumType.STRING)
    @Column(name = "card_network")
    private CardNetwork cardNetwork;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private Status status;

    @Column(name = "expiry_date")
    private LocalDate expiryDate;

    @Column(name = "credit_limit", precision = 15, scale = 2)
    private BigDecimal creditLimit;

    @Column(name = "current_balance", precision = 15, scale = 2)
    private BigDecimal currentBalance;

    @Column(name = "cashback_earned", precision = 15, scale = 2)
    private BigDecimal cashbackEarned;

    @Column(name = "daily_limit", precision = 15, scale = 2)
    private BigDecimal dailyLimit;

    @Column(name = "monthly_limit", precision = 15, scale = 2)
    private BigDecimal monthlyLimit;

    @Column(name = "is_international_enabled")
    private Boolean isInternationalEnabled;

    @Column(name = "is_online_enabled")
    private Boolean isOnlineEnabled;

    @Column(name = "is_contactless_enabled")
    private Boolean isContactlessEnabled;

    @Column(name = "applied_by")
    private UUID appliedBy;

    @Column(name = "approved_by")
    private UUID approvedBy;

    @Column(name = "approved_at")
    private LocalDateTime approvedAt;

    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @PrePersist
    protected void onCreate() {
        createdAt = LocalDateTime.now();
        if (currentBalance == null) currentBalance = BigDecimal.ZERO;
        if (cashbackEarned == null) cashbackEarned = BigDecimal.ZERO;
        if (status == null) status = Status.PENDING;
        if (dailyLimit == null) dailyLimit = new BigDecimal("50000.00");
        if (monthlyLimit == null) monthlyLimit = new BigDecimal("500000.00");
        if (isInternationalEnabled == null) isInternationalEnabled = false;
        if (isOnlineEnabled == null) isOnlineEnabled = true;
        if (isContactlessEnabled == null) isContactlessEnabled = true;
        if (expiryDate == null) expiryDate = LocalDate.now().plusYears(5);
    }

    public enum CardType { DEBIT, CREDIT, PREPAID, FLEXI }
    public enum CardNetwork { VISA, MASTERCARD, RUPAY }
    public enum Status { PENDING, ACTIVE, FROZEN, CANCELLED, REJECTED }
}

File: backend-python/com/gxs/bank/model/Card.py
from datetime import date, datetime
from decimal import Decimal
from enum import Enum
import uuid

from sqlalchemy import Boolean, Column, Date, DateTime, Enum as SQLEnum, ForeignKey, Numeric, String
from sqlalchemy.orm import relationship

from com.gxs.bank.config.database import Base


class CardType(str, Enum):
    DEBIT = "DEBIT"
    CREDIT = "CREDIT"
    PREPAID = "PREPAID"
    FLEXI = "FLEXI"


class CardNetwork(str, Enum):
    VISA = "VISA"
    MASTERCARD = "MASTERCARD"
    RUPAY = "RUPAY"


class Status(str, Enum):
    PENDING = "PENDING"
    ACTIVE = "ACTIVE"
    FROZEN = "FROZEN"
    CANCELLED = "CANCELLED"
    REJECTED = "REJECTED"


class Card(Base):
    __tablename__ = "cards"

    id = Column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    userId = Column("user_id", String(36), ForeignKey("users.id"), nullable=False)
    user = relationship("User", lazy="select")

    cardNumberLast4 = Column("card_number_last4", String(4), nullable=False)
    cardHolderName = Column("card_holder_name", String(255), nullable=True)
    cardType = Column("card_type", SQLEnum(CardType), nullable=False)
    cardNetwork = Column("card_network", SQLEnum(CardNetwork), nullable=True)
    status = Column(SQLEnum(Status), nullable=False, default=Status.PENDING)
    expiryDate = Column("expiry_date", Date, nullable=False, default=lambda: date.today().replace(year=date.today().year + 5))
    creditLimit = Column("credit_limit", Numeric(15, 2), nullable=True)
    currentBalance = Column("current_balance", Numeric(15, 2), nullable=False, default=Decimal("0.00"))
    cashbackEarned = Column("cashback_earned", Numeric(15, 2), nullable=False, default=Decimal("0.00"))
    dailyLimit = Column("daily_limit", Numeric(15, 2), nullable=False, default=Decimal("50000.00"))
    monthlyLimit = Column("monthly_limit", Numeric(15, 2), nullable=False, default=Decimal("500000.00"))
    isInternationalEnabled = Column("is_international_enabled", Boolean, nullable=False, default=False)
    isOnlineEnabled = Column("is_online_enabled", Boolean, nullable=False, default=True)
    isContactlessEnabled = Column("is_contactless_enabled", Boolean, nullable=False, default=True)
    appliedBy = Column("applied_by", String(36), nullable=True)
    approvedBy = Column("approved_by", String(36), nullable=True)
    approvedAt = Column("approved_at", DateTime, nullable=True)
    createdAt = Column("created_at", DateTime, nullable=False, default=datetime.utcnow)


Card.CardType = CardType
Card.CardNetwork = CardNetwork
Card.Status = Status
